### ISSUE TITLE
Evidence file is not created

### ISSUE THREAD

**Issue by richelbilderbeek** [richelbilderbeek/pirouette#111]
See [this Travis CI log](https://travis-ci.org/richelbilderbeek/pirouette/builds/491999942#L1842):

```
 ERROR
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
  [1] "ncol(df): 13"
  [1] "length(nltts): 3"
  ── 1. Error: generative with twin (@test-pir_run.R#759)  ───────────────────────
  all(file.exists(filenames)) is not TRUE
  1: testit::assert(all(file.exists(filenames))) at testthat/test-pir_run.R:759
  2: assert2(fact, if (one) mc[[2]][-1] else mc[-1], parent.frame(), !one)
  3: stop(sprintf(ngettext(length(val), "%s is not TRUE", "%s are not all TRUE"), deparse_key(expr)), 
         call. = FALSE, domain = NA)
  
  ══ testthat results  ═════════════════════════════════════════���═════════════════
  OK: 502 SKIPPED: 2 FAILED: 1
  1. Error: generative with twin (@test-pir_run.R#759) 
```

**Comment by richelbilderbeek** [comment]
The evidence file is somehow not created:

```r
test_that("generative with twin", {

  if (!beastier::is_on_travis()) return()
  skip("Issue 111, #111")

  # ...

  experiment <- create_experiment(
    model_type = "generative",
    run_if = "always",
    do_measure_evidence = FALSE,
    # ...
  )
  # ...
  filenames <- c(
    pir_params$alignment_params$fasta_filename,
    pir_params$evidence_filename,
    # ...
  )

  errors <- pir_run(
    phylogeny = phylogeny,
    pir_params = pir_params
  )

  # Files created
  testit::assert(file.exists(filenames[2])) # test fails here
```

**Comment by richelbilderbeek** [comment]
Doubly weird, in the PDF it works:

[twinning.pdf](https://github.com/richelbilderbeek/pirouette/files/2854809/twinning.pdf)

A picture of it:

![2019-02-12-093717_1398x1056_scrot](https://user-images.githubusercontent.com/2098230/52622371-f8cc4c80-2ea9-11e9-8489-95b7174649e7.png)

**Comment by richelbilderbeek** [comment]
I think @Giappo will like this one. Good luck :male_detective:

**Comment by Giappo** [comment]
@richelbilderbeek there is something I don't understand. I'll try to explain how I see the problem and you can tell me where am I wrong or if instead I am right.

We are testing for the "generative" type.
This means that the evidence is, in principle, not needed because the model we are going to use is always the "generative" one. If I try to dig into the functions, starting from the test you were pointing at I arrive here:
"pir_run.R" R.124-132:
  # Estimate evidences (aka marginal likelihoods) if needed
  # marg_liks will be NULL if this was unneeded, for example, when
  # interested in the generative model only
  marg_liks <- est_evidences(
    fasta_filename = alignment_params$fasta_filename,
    experiments = experiments,
    evidence_filename = evidence_filename,
    verbose = verbose
  )

If I run these lines, as foretold by the comment, I get this:

> marg_liks
NULL

When I arrive at R.216 I find this:

# Add evidence (marginal likelihoods) in columns
  if (!is.null(marg_liks)) {
    for (i in seq_along(experiments)) {
      experiment <- experiments[[i]]
      marg_liks_row <- which(
        marg_liks$site_model_name ==
          experiment$inference_model$site_model$name &
        marg_liks$clock_model_name ==
          experiment$inference_model$clock_model$name &
        marg_liks$tree_prior_name ==
          experiment$inference_model$tree_prior$name
      )
      # if there is no row, 'which' returns a zero-length vector
      # Happens when the generative model is not part of the models
      # under selection
      if (length(marg_liks_row) != 0) {
        df$inference_model_weight[i] <- marg_liks$weight[marg_liks_row]
      }
    }
  }

The if statement is not triggered because marg_liks is NULL.

So, two questions:
1) I wanted to ask you in what part of this pipeline (pir_run -> pir_run_tree -> phylo_to_errors) the files are supposed to be created (it's not easy to spot). Despite my ignorance on that, I don't see how a file can be even created, since there is no data on evidence (because marg_liks == NULL).
2) Are we sure we even want an evidence file to be created where model_type == "generative"? If yes, what information is it supposed to contain?

**Comment by Giappo** [comment]
> Doubly weird, in the PDF it works:
> 
> [twinning.pdf](https://github.com/richelbilderbeek/pirouette/files/2854809/twinning.pdf)
> 
> A picture of it:
> 
> ![2019-02-12-093717_1398x1056_scrot](https://user-images.githubusercontent.com/2098230/52622371-f8cc4c80-2ea9-11e9-8489-95b7174649e7.png)

Are you sure you're using the "generative" model_type here?

**Comment by richelbilderbeek** [comment]
Dumb me, when only using only models with `measure_evidence == FALSE`, there is no marginal likelihood file created, as it need not be calculated :blush: !

**Comment by richelbilderbeek** [comment]
> I wanted to ask you in what part of this pipeline (pir_run -> pir_run_tree -> phylo_to_errors) the files are supposed to be created (it's not easy to spot). Despite my ignorance on that, I don't see how a file can be even created, since there is no data on evidence (because marg_liks == NULL).

Agreed, when no `estimate_evidence == TRUE`, there will be no evidences estimated.

> Are we sure we even want an evidence file to be created where model_type == "generative"? If yes, what information is it supposed to contain?

Agreed, we don't want such a file to be created.

> Are you sure you're using the "generative" model_type here?

The vignette uses `do_measure_evidence` for a best candidate model:

```
create_experiment(
  do_measure_evidence = TRUE,
  # ...
)
```

Sorry for the noise.

**Comment by Giappo** [comment]
It's fine. I used this time to understand more how pirouette works. It's been useful to me.
You can close this issue if you think there is nothing more to do.
Otherwise, explain me what you want to do more.
:octopus:

**Comment by richelbilderbeek** [comment]
Cool, will close this Issue :octopus:

**Comment by Giappo** [comment]
Even though when using only models with `do_measure_evidence == FALSE`, there is no marginal likelihood file created, there are still tests in "test-pir_run" that require to be fixed.
I think that it would be easy for @richelbilderbeek to put them in the right way.
I need this to close #114.
To find the tests that need to be fixed look for #111 in the code. There should be two occurrences.

**Comment by richelbilderbeek** [comment]
Thanks for the assign, will do so now :+1:
